bftengine/src/bftengine/ReplicaImp.hpp

```diff
@@ -120,8 +120,17 @@ class ReplicaImp : public InternalReplicaApi, public ReplicaForStateTransfer {
   size_t NonPrimaryCombinedReqSize = 1000;
   //
   const std::thread::id MAIN_THREAD_ID;
+  // A preprepare message which is still building is called transient preprepare.
+  // activeTransientPreprepare_ represents the number of preprepare messages that are
+  // currently getting build in a separate thread.
+  // This should not be thread safe because its used only in the main thread
+  // This should start with 0, since by default we will assume that a preprepare is taken
+  // to be added to main log. And this will solely represent those preprepares which are
+  // to be added to main log, but are not added as yet.
+  uint16_t numOfTransientPreprepareMsgs_ = 0;
```

Comment:
I think this variable should be initialized also when state transfer is completed.

Reply:
I'm thinking about what can happen on a view change if, let's say, replica 0 is the primary. A VC happens when some PP is inflight. Can this variable be on a non 0 state after the VC? If it becomes the primary again, will this value be dirty?

Reply:
Thanks @cloudnoize for reviewing and raising this question. Let me discuss the interaction with view change, State Transfer and Pruning and effect on this variable here.
How is view change handled?
Ans:
Inside [finishAddingRequestsToPrePrepareMsg()](https://github.com/arc-vmware/concord-bft/blob/25960bdadfb803ba0bda7510aa9361935f7c36d0/bftengine/src/bftengine/ReplicaImp.cpp#L742) function, a thread is created to compute the digest of digests for the given preprepare message. The digest of digests is a time consuming process (it might even take upto 2 ms).
So lets assume that there are ‘x’ preprepare messages in transient state.
 
If any of the preprepare message is done with this computation, it is sent as an internal message (in any order). When this message comes to the main thread it is handled in the [onInternalMsg()](https://github.com/arc-vmware/concord-bft/blob/25960bdadfb803ba0bda7510aa9361935f7c36d0/bftengine/src/bftengine/ReplicaImp.cpp#L1536) handler and it calls the [startConsensusProcess()](https://github.com/arc-vmware/concord-bft/blob/25960bdadfb803ba0bda7510aa9361935f7c36d0/bftengine/src/bftengine/ReplicaImp.cpp#L1563) function. The first line of startConsensusProcess() function calls [isCurrentPrimary()](https://github.com/arc-vmware/concord-bft/blob/25960bdadfb803ba0bda7510aa9361935f7c36d0/bftengine/src/bftengine/ReplicaImp.cpp#L865) and then takes a position in the main log of that replica. If in the meantime a view change happens and it’s no longer a primary, then the preprepare is simply dropped (and a memory leak might happen, which I will fix in 1.6 for the memory leak fix for preprepare action item, which we had discussed earlier).
 
So the view change is actually handled in this scenario.

How ST and Pruning is handled?
The pruning and ST are also handled as below:
Inside [finishAddingRequestsToPrePrepareMsg()](https://github.com/arc-vmware/concord-bft/blob/25960bdadfb803ba0bda7510aa9361935f7c36d0/bftengine/src/bftengine/ReplicaImp.cpp#L742) function, the ST is checked as a part of the prerequisite. Assume that during the handling of preprepare digest of digests the state transfer starts. When we receive the message in the onInternalMsg() handler we [check again for both pruning and ST](https://github.com/arc-vmware/concord-bft/blob/25960bdadfb803ba0bda7510aa9361935f7c36d0/bftengine/src/bftengine/ReplicaImp.cpp#L1557). We will not start consensus and delete the message.
 
So this way both pruning and ST are handled together. I will setup a call today, whenever you have time and discuss the above.

Can this variable become dirty?
There are 3 regions of this variable:
a) negative region -- This will never happen as this is an unsigned variable. I will add a check to reduce this variable only when its positive. But this variable is never reduced more than required. Also this variable is always decremented in the message thread and none of the operations on this variable is happening in parallel. 
b) The count of transient PPM : Desired state and its fine.
c) Its incremented more than the count of transient ppms. This is not possible as its incremented only when we are creating a new thread for computation. And its decrementing only when we are receiving the result of computation. This entire process is happening is same machine, so this variable will become zero.